collections/_tutorials/2022-06-20-tutorial_transformer.md

```diff
@@ -0,0 +1,124 @@
+---
+layout: post
+title:  "The transformer paradigm demystified"
+author: 'Olivier Bernard'
+date:   2022-06-20
+categories: transformer, encoder
+---
+
+# Notes
+
+* Here is a link to an interesting video to understand the transformer theory: [video](https://www.youtube.com/watch?v=YAgjfMR9R_M)
+
+&nbsp;
+
+- [**Introduction**](#introduction)
+- [**Tokenization process**](#tokenization-process)
+- [**Transformer encoder**](#transformer-encoder)
+  - [Norm step](#norm-step)
+  - [MLP step](#mlp-step)  
+  - [Multi-Head Attention block](#multi-head-attention-block)  
+  - [Self-Attention module](#self-attention-module)    
+
+&nbsp;
+
+## **Introduction**
+Transformer paradigm has been first successfully applied in Natural Langage Processing (NLP) and is currently highly investigated for image processing. 
+
+The basic concept of transformer lies in the generation of self attention maps to guide the decision making process of the underlying architecture. To illsutration the underlying mechanism, we will open the black box of the famous [Vision Transformer (ViT)](https://creatis-myriad.github.io/2022/06/01/VisionTransformer.html) method. The overall architecture is given below:
+
+![](/collections/images/transformers/vit_overview.jpg)
+
+&nbsp;
+
+## **Tokenization process**
+The first step is to convert the input data into a sequence of tokens, $$i.e.$$ a sequence of vectors of dimensions $$\mathbb{R}^{1 \times (P^2 C)}$$, with $$P^2 C$$ being the length of each flattened patch. This step is called the tokenization process and involves a simple linear projection ($$i.e.$$ a multiplication with a matrix of dimensions $$\mathbb{R}^{(P^2 C) \times D}$$) and a position embedding step which encodes spatial information. The same linear projection and position embedding are shared to encode each patch. This process is modeled as follows: 
```

Comment:
I would specify that it's the same position embedding **operation** that is shared, otherwise it could be interpreted as if all patches have the same embedding.
```suggestion
The first step is to convert the input data into a sequence of tokens, *i.e.* a sequence of vectors of dimensions $$\mathbb{R}^{1 \times (P^2 C)}$$, with $$P^2 C$$ being the length of each flattened patch. This step is called the tokenization process and involves a simple linear projection (*i.e.* a multiplication with a matrix of dimensions $$\mathbb{R}^{(P^2 C) \times D}$$) and a position embedding step which encodes spatial information. The same linear projection and position embedding operation are shared to encode each patch. This process is modeled as follows: 
```